# MySQL database

### <span style='color:Yellow'> 1. Create a script to connection with database.  </span>

In [3]:
from sqlalchemy import create_engine, MetaData, Table

db_string = "mysql://bauer1:W0JjW9jge2SYwwYh@mysql.agh.edu.pl:3306/bauer1"

db = create_engine(db_string)

print(db.table_names())

C:\Users\User\AppData\Local\Temp/ipykernel_1348/3525206458.py:7: SADeprecationWarning: The Engine.table_names() method is deprecated and will be removed in a future release.  Please refer to Inspector.get_table_names(). (deprecated since: 1.4)
  print(db.table_names())


['bookings', 'cities', 'countries', 'customers', 'employees', 'hosts', 'offices', 'orderdetails', 'orders', 'payments', 'places', 'productlines', 'products', 'reviews', 'users']


### <span style='color:Yellow'> 2. Based on information_schema, present how to explore the relationships between the tables: </span>

 The way to explore the relationships between the tables is to check the primary key and foreign key of each columns in both tables. A primary key is used to ensure data in the specific column is unique. A foreign key is a column or group of columns in a relational database table that provides a link between data in two tables. 

### <span style='color:Lime'> i) customers and employees

In [4]:
metadata = MetaData()
table_customers = Table('customers', metadata , autoload=True, autoload_with=db)
table_employees = Table('employees', metadata , autoload=True, autoload_with=db)

print(' Customers: ')
print(repr(table_customers))
print(table_customers.columns.keys())

print('\n Employees: ')
print(repr(table_employees))
print(table_employees.columns.keys())

 Customers: 
Table('customers', MetaData(), Column('customerNumber', INTEGER(display_width=11), table=<customers>, primary_key=True, nullable=False), Column('customerName', VARCHAR(length=50), table=<customers>, nullable=False), Column('contactLastName', VARCHAR(length=50), table=<customers>, nullable=False), Column('contactFirstName', VARCHAR(length=50), table=<customers>, nullable=False), Column('phone', VARCHAR(length=50), table=<customers>, nullable=False), Column('addressLine1', VARCHAR(length=50), table=<customers>, nullable=False), Column('addressLine2', VARCHAR(length=50), table=<customers>), Column('city', VARCHAR(length=50), table=<customers>, nullable=False), Column('state', VARCHAR(length=50), table=<customers>), Column('postalCode', VARCHAR(length=15), table=<customers>), Column('country', VARCHAR(length=50), table=<customers>, nullable=False), Column('salesRepEmployeeNumber', INTEGER(display_width=11), ForeignKey('employees.employeeNumber'), table=<customers>), Column('cr

### <span style='color:red'> Result: </span>

We can see that in the Customers table there is one column 'salesRepEmployeeNumber' has ForeignKey to the Employees table column 'employeeNumber'. And in the Employees table has two column 'officeCode' and 'reportsTo' have ForeignKey. Therefore, we can there are relationships between them.

### <span style='color:Lime'> ii) customers, payments and orders

In [5]:
table_payments = Table('payments', metadata , autoload=True, autoload_with=db)
table_orders= Table('orders', metadata , autoload=True, autoload_with=db)

print('\n Payments: ')
print(repr(table_payments))
print(table_payments.columns.keys())

print('\n Orders: ')
print(repr(table_orders))
print(table_orders.columns.keys())


 Payments: 
Table('payments', MetaData(), Column('customerNumber', INTEGER(display_width=11), ForeignKey('customers.customerNumber'), table=<payments>, primary_key=True, nullable=False), Column('checkNumber', VARCHAR(length=50), table=<payments>, primary_key=True, nullable=False), Column('paymentDate', DATE(), table=<payments>, nullable=False), Column('amount', DECIMAL(precision=10, scale=2), table=<payments>, nullable=False), schema=None)
['customerNumber', 'checkNumber', 'paymentDate', 'amount']

 Orders: 
Table('orders', MetaData(), Column('orderNumber', INTEGER(display_width=11), table=<orders>, primary_key=True, nullable=False), Column('orderDate', DATE(), table=<orders>, nullable=False), Column('requiredDate', DATE(), table=<orders>, nullable=False), Column('shippedDate', DATE(), table=<orders>), Column('status', VARCHAR(length=15), table=<orders>, nullable=False), Column('comments', TEXT(), table=<orders>), Column('customerNumber', INTEGER(display_width=11), ForeignKey('custome

### <span style='color:red'> Result: </span>

We can see that in the Payments table there is one column 'customerNumber' has ForeignKey to the Customers table column 'customerNumber'. And the same goes for the Employees table. Therefore, we can there are relationships between three of them.

### <span style='color:Yellow'> 3. How many products are in the store? </span>

In [7]:
from sqlalchemy import select

table_products = Table('products', metadata , autoload=True, autoload_with=db)
stmt = 'select count(*) from products'
print(stmt)
results = db.execute(stmt).fetchall()
results

select distinct count(*) from products


[(110,)]

### <span style='color:red'> Result: </span> 110 product

### <span style='color:Yellow'> 4. Display list of offices with limit 5. </span> 

In [9]:
table_offices = Table('offices', metadata , autoload=True, autoload_with=db)
stmt = select([table_offices])
print(stmt)
results = db.execute(stmt).fetchmany(size=5)
results

SELECT offices."officeCode", offices.city, offices.phone, offices."addressLine1", offices."addressLine2", offices.state, offices.country, offices."postalCode", offices.territory 
FROM offices


[('1', 'San Francisco', '+1 650 219 4782', '100 Market Street', 'Suite 300', 'CA', 'USA', '94080', 'NA'),
 ('2', 'Boston', '+1 215 837 0825', '1550 Court Place', 'Suite 102', 'MA', 'USA', '02107', 'NA'),
 ('3', 'NYC', '+1 212 555 3000', '523 East 53rd Street', 'apt. 5A', 'NY', 'USA', '10022', 'NA'),
 ('4', 'Paris', '+33 14 723 4404', "43 Rue Jouffroy D'abbans", None, None, 'France', '75017', 'EMEA'),
 ('5', 'Tokyo', '+81 33 224 5000', '4-1 Kioicho', None, 'Chiyoda-Ku', 'Japan', '102-8578', 'Japan')]

### <span style='color:Yellow'> 5. Find the oldest and youngest payments in rental. </span> 

In [54]:
table_payments = Table('payments', metadata , autoload=True, autoload_with=db)
stmt = 'select min(paymentDate) from payments'
results = db.execute(stmt).fetchall()
print(stmt)
results

select min(paymentDate) from payments


[(datetime.date(2003, 1, 16),)]

In [55]:
stmt = 'select max(paymentDate) from payments'
results = db.execute(stmt).fetchall()
print(stmt)
results

select max(paymentDate) from payments


[(datetime.date(2005, 6, 9),)]

### <span style='color:red'> Result: </span> 
Earliest payment (2003-01-16)  
Latest payment (2005-06-09)
